operator/validator/controller.go

```diff
@@ -436,43 +436,59 @@ func (c *controller) handleNonCommitteeMessages(
+// StartValidators loads all persisted shares and sets up the corresponding validators
+func (c *controller) StartValidators(ctx context.Context) error {
 	// TODO: Pass context whereever the execution flow may be blocked.
 
-	// Load non-liquidated shares.
-	shares := c.sharesStorage.List(nil, registrystorage.ByNotLiquidated())
-	if len(shares) == 0 {
+	if c.validatorCommonOpts.ExporterOptions.Enabled {
+		// There are no committee validators to set up.
 		close(c.committeeValidatorSetup)
-		c.logger.Info("could not find validators")
-		return
+		return nil
 	}
 
-	var ownShares []*ssvtypes.SSVShare
-	for _, share := range shares {
-		if c.operatorDataStore.GetOperatorID() != 0 && share.BelongsToOperator(c.operatorDataStore.GetOperatorID()) {
-			ownShares = append(ownShares, share)
+	init := func() ([]*validator.Validator, []*validator.Committee, error) {
+		defer close(c.committeeValidatorSetup)
+
+		// Load non-liquidated shares that belong to our own Operator.
+		ownShares := c.sharesStorage.List(
+			nil,
+			registrystorage.ByNotLiquidated(),
+			registrystorage.ByOperatorID(c.operatorDataStore.GetOperatorID()),
+		)
+		if len(ownShares) == 0 {
+			c.logger.Info("no own non-liquidated validator shares found in DB")
+			return nil, nil, nil
 		}
-	}
 
-	if c.validatorCommonOpts.ExporterOptions.Enabled {
-		// There are no committee validators to setup.
-		close(c.committeeValidatorSetup)
-	} else {
 		// Setup committee validators.
-		inited, committees := c.setupValidators(ownShares)
-		if len(inited) == 0 {
-			// If no validators were started and therefore we're not subscribed to any subnets,
-			// then subscribe to a random subnet to participate in the network.
-			if err := c.network.SubscribeRandoms(1); err != nil {
-				c.logger.Error("failed to subscribe to random subnets", zap.Error(err))
-			}
+		validators, committees := c.setupValidators(ownShares)
+		if len(validators) == 0 {
+			return nil, nil, fmt.Errorf("none of %d validators were successfully initialized", len(ownShares))
 		}
-		close(c.committeeValidatorSetup)
 
-		// Start validators.
-		c.startValidators(inited, committees)
+		return validators, committees, nil
+	}
+
+	// Initialize validators.
+	validators, committees, err := init()
+	if err != nil {
+		return fmt.Errorf("init validators: %w", err)
+	}
+	if len(validators) == 0 {
+		// If no validators were initialized - we're not subscribed to any subnets,
+		// we have to subscribe to 1 random subnet to participate in the network.
+		if err := c.network.SubscribeRandoms(1); err != nil {
+			c.logger.Error("failed to subscribe to random subnets", zap.Error(err))
+		}
+		return nil
+	}
+
+	// Start validators.
+	started := c.startValidators(validators, committees)
+	if started == 0 {
+		return fmt.Errorf("none of %d validators were successfully started", len(validators))
 	}
+	return nil
 }
 
 // setupValidators setup and starts validators from the given shares.
```

Comment:
not related to your changes

The comment is misleading. This function only sets up validators, it doesn't start them.
// setupValidators sets up validators from the given shares.